README.md
# Native UI Select Angular component
## ng2-select [![npm version](https://badge.fury.io/js/ng2-select-hoveytech.svg)](http://badge.fury.io/js/ng2-select-hoveytech) [![npm downloads](https://img.shields.io/npm/dm/ng2-select-hoveytech.svg)](https://npmjs.org/ng2-select-hoveytech)[![slack](https://ngx-slack.herokuapp.com/badge.svg)](https://ngx-slack.herokuapp.com)

[![Build Status](https://travis-ci.org/hoveytech/ng2-select.svg?branch=development)](https://travis-ci.org/hoveytech/ng2-select-hoveytech)

#### <span style="color: blue;">NOTE: This is a port of ng2-select library by [valor-software](https://github.com/valor-software). due to this [issue](https://github.com/valor-software/ng2-select/issues/769).</span>

## Quick start

1. A recommended way to install ***ng2-select-hoveytech*** is through [npm](https://www.npmjs.com/search?q=ng2-select-hoveytech) package manager using the following command:

  `npm i ng2-select-hoveytech --save`

## API

### Properties

  - `items` - (`Array<any>`) - Array of items from which to select. Should be an array of objects with `id` and `text` properties.
  As convenience, you may also pass an array of strings, in which case the same string is used for both the ID and the text.
  Items may be nested by adding a `children` property to any item, whose value should be another array of items. Items that have children may omit having an ID.
  If `items` are specified, all items are expected to be available locally and all selection operations operate on this local array only.
  If omitted, items are not available locally, and the `query` option should be provided to fetch data.
  - `active` (`?Array<any>`) - selection data to set. This should be an object with `id` and `text` properties in the case of input type 'Single',
  or an array of such objects otherwise. This option is mutually exclusive with value.
  - `allowClear` (`?boolean=false`) (*not yet supported*) - Set to `true` to allow the selection to be cleared. This option only applies to single-value inputs.
  - `placeholder` (`?string=''`) - Placeholder text to display when the element has no focus and selected items.
  - `disabled` (`?boolean=false`) - When `true`, it specifies that the component should be disabled.
  - `multiple` - (`?boolean=false`) - Mode of this component. If set `true` user can select more than one option.
  - `dropdown` (`?boolean=true`) - Set to false if you want the dropup class applied.

  This option only applies to single-value inputs, as multiple-value inputs don't have the search input in the dropdown to begin with.

### Events

  - `data` - it fires during all events of this component; returns `Array<any>` - current selected data
  - `selected` - it fires after a new option selected; returns object with `id` and `text` properties that describes a new option.
  - `removed` - it fires after an option removed; returns object with `id` and `text` properties that describes a removed option.
  - `typed` - it fires after changing of search input; returns `string` with that value.

# Troubleshooting

Please follow this guidelines when reporting bugs and feature requests:

1. Use [GitHub Issues](https://github.com/hoveytech/ng2-select/issues) board to report bugs and feature requests (not our email address)
2. Please **always** write steps to reproduce the error. That way we can focus on fixing the bug, not scratching our heads trying to reproduce it.

Thanks for understanding!

### License

The MIT License (see the [LICENSE](https://github.com/hoveytech/ng2-select/blob/master/LICENSE) file for the full text)


CHANGELOG.md
<a name="0.1.6"></a>
## [0.1.6](https://github.com/hoveytech/ng2-select/compare/v0.1.5...v0.1.6) (2017-05-02)

### Bug fixes

* Fixed issue with setting items async, while options are displayed.

<a name="0.1.5"></a>
## [0.1.5](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.5) (2017-04-26)

### Bug fixes

* Silly bug fix for disabled modification using FormControl

<a name="0.1.4"></a>
## [0.1.4](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.4) (2017-04-26)

### Enhancement

* Added support for FormControl setting disabled for control

<a name="0.1.3"></a>
## [0.1.3](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.3) (2017-04-26)

### Bug fixes

* Fix for active property setting ngModel changed callback

<a name="0.1.2"></a>
## [0.1.2](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.2) (2017-04-25)

### Bug fixes

* Fix to active property setter for calling change callback

<a name="0.1.1"></a>
## [0.1.1](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.1) (2017-04-24)

### Features

* Added 'dropdown' @Input.

<a name="0.1.0"></a>
# [0.1.0](https://github.com/hoveytech/ng2-select/compare/v1.2.0...v0.1.0) (2017-04-24)

### Bug Fixes

* Fixed issue with select component not calling ngModel callbacks for touched and value changed.

### Features

* Added 'closed' EventEmitter to select component.


scripts/test.ts
import '../demo/src/polyfills.ts';

import 'zone.js/dist/long-stack-trace-zone';
import 'zone.js/dist/proxy.js';
import 'zone.js/dist/sync-test';
import 'zone.js/dist/jasmine-patch';
import 'zone.js/dist/async-test';
import 'zone.js/dist/fake-async-test';
import { getTestBed } from '@angular/core/testing';
import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting
} from '@angular/platform-browser-dynamic/testing';
import '@angular/platform-browser/testing';

import './matchers';

// Unfortunately there's no typing for the `__karma__` variable. Just declare it as any.
declare var __karma__: any;
declare var require: any;

// Prevent Karma from running prematurely.
__karma__.loaded = Function.prototype;

// First, initialize the Angular testing environment.
getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting()
);
// Then we find all the tests.
let context = require.context('../demo/src', true, /\.spec\.ts/);
// And load the modules.
context.keys().map(context);

let context2 = require.context('../src/spec', true, /\.spec\.ts/);
context2.keys().map(context2);
// Finally, start Karma to run the tests.
__karma__.start();


src/spec/select.spec.ts
import { Component, OnInit, Type, DebugElement } from '@angular/core';
import { ComponentFixture, TestBed, fakeAsync, tick } from '@angular/core/testing';

import { SelectModule, SelectComponent } from '../ng2-select';
import { SelectItem } from "../index";
import { FormBuilder, FormControl, Validators, FormGroup, FormsModule, ReactiveFormsModule } from "@angular/forms";

const html = ``;

describe('Component: ng2-select', () => {

    beforeEach(() => {
        TestBed.configureTestingModule({
            declarations: [TestSelectEmptyComponent, TestSelectActiveComponent, TestSelectAsyncComponent],
            imports: [SelectModule, FormsModule, ReactiveFormsModule]
        });
        TestBed.overrideComponent(TestSelectEmptyComponent, { set: { template: html } });
    });

    let initializeFixture = <TFixture>(fixtureType: Type<TFixture>, template: string) : ComponentFixture<TFixture> => { 
        TestBed.overrideComponent(fixtureType, { set: { template } });
        let fixture = TestBed.createComponent(fixtureType);
        let context = fixture.componentInstance;
        fixture.detectChanges();
        tick();
        fixture.detectChanges();
        return fixture;
    };

    it('fixture should not be null', fakeAsync(() => {
        TestBed.overrideComponent(TestSelectActiveComponent, { set: { template: '<div></div>' } });
        let fixture = TestBed.createComponent(TestSelectActiveComponent);
        let context = fixture.componentInstance;
        fixture.detectChanges();
        tick();

        expect(fixture).not.toBeNull();
    }));

    it('does set implement ngModel and sets active internally if using ngModel', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<form [formGroup]="formGroup"><ng-select formControlName="select" [ngModel]="selected" [items]="items"></ng-select></form>');
        let comp = fixture.componentInstance;

        expect(comp.select.value).toBeDefined();
        expect(comp.select.value.length).toBe(1);
        expect(comp.select.value[0].id).toBe(1);
        expect(comp.select.errors).toBeNull();
        expect(comp.select.valid).toBeTruthy();
        expect(comp.select.pristine).toBeTruthy();
        expect(comp.select.touched).toBeFalsy();

        let selectEl = fixture.debugElement.children[0].children[0];
        let selectComponent = <SelectComponent>selectEl.injector.get(SelectComponent);

        expect(selectComponent.active).toBeDefined();
        expect(selectComponent.active.length).toBe(1);
        expect(selectComponent.active[0].id).toBe(1);
    }));

    it('does set disabled if disable called using formControl', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<form [formGroup]="formGroup"><ng-select formControlName="select" [ngModel]="selected" [items]="items"></ng-select></form>');
        let comp = fixture.componentInstance;
        let selectEl = fixture.debugElement.children[0].children[0];
        let selectComponent = <SelectComponent>selectEl.injector.get(SelectComponent);

        expect(selectComponent.disabled).toBeFalsy();
        comp.select.disable();
        expect(selectComponent.disabled).toBeTruthy();
    }));

    it('does set enabled if enabled called using formControl', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<form [formGroup]="formGroup"><ng-select formControlName="select" [ngModel]="selected" [items]="items"></ng-select></form>');
        let comp = fixture.componentInstance;
        let selectEl = fixture.debugElement.children[0].children[0];
        let selectComponent = <SelectComponent>selectEl.injector.get(SelectComponent);

        comp.select.disable();
        comp.select.enable();
        expect(selectComponent.disabled).toBeFalsy();
    }));

    let openOptions = ( ngSelectElement: DebugElement) => { 
        var buttonEl = ngSelectElement.children[0].children[1].children[0];
        buttonEl.nativeElement.click();
    }

    let openAndCloseOptions = ( ngSelectElement: DebugElement) => { 
        openOptions(ngSelectElement);
        ngSelectElement.parent.nativeElement.click();
        tick();
    }

    it('does set call ngModel touched callback if dropdown opened and closed', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<form [formGroup]="formGroup"><ng-select formControlName="select" ngModel [active]="selected" [items]="items" ></ng-select></form>');
        let comp = fixture.componentInstance;

       openAndCloseOptions(fixture.debugElement.children[0].children[0]);
        
        expect(comp.select.touched).toBeTruthy();
    }));

    it('does emit on closed EventEmitter', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<form [formGroup]="formGroup"><ng-select formControlName="select" ngModel [active]="selected" [items]="items" ></ng-select></form>');
        let comp = fixture.componentInstance;

        let selectEl = fixture.debugElement.children[0].children[0];
        let select = <SelectComponent>selectEl.injector.get(SelectComponent);

        let called: boolean = false;
        select.closed.subscribe(() => {
            called = true;
        })
        openAndCloseOptions(selectEl);
        
        expect(called).toBeTruthy();
    }));

    it('does apply class dropdown cssClass by default', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<ng-select [active]="selected" [items]="items"></ng-select>');
        let comp = fixture.componentInstance;

        let selectEl = fixture.debugElement.children[0].children[0];
        expect(selectEl.nativeElement.classList.contains('dropdown')).toBeTruthy();
        expect(selectEl.nativeElement.classList.contains('dropup')).toBeFalsy();
    }));

    it('does apply class dropup when dropdown is false', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectActiveComponent, 
             '<ng-select [dropdown]="false" [active]="selected" [items]="items"></ng-select>');
        let comp = fixture.componentInstance;

        let selectEl = fixture.debugElement.children[0].children[0];
        expect(selectEl.nativeElement.classList.contains('dropup')).toBeTruthy();
        expect(selectEl.nativeElement.classList.contains('dropdown')).toBeFalsy();
    }));

    it('does display options when items added after typing', fakeAsync(() => {
        let fixture = initializeFixture(TestSelectAsyncComponent, 
             '<ng-select (typed)="onFilterChanged($event)" [items]="items"></ng-select>');
        let comp = fixture.componentInstance;

        let selectEl = fixture.debugElement.children[0];
        let select = <SelectComponent>selectEl.injector.get(SelectComponent);
        openOptions(selectEl);
        fixture.detectChanges();
        let inputEl = selectEl.children[0].children[1];
        inputEl.nativeElement.value = 'T';
        inputEl.nativeElement.dispatchEvent(new KeyboardEvent('keydown',{
            "key": "T"
        }));

        tick();
        fixture.detectChanges();

        let optionsUlEl = selectEl.children[0].children[2];
        expect(optionsUlEl.children.length).toBe(3); // count of items in options.
    }));
});

class CustomSelectItemObject {
    constructor(public name?: string, public identifier?: string) {
    }
}

@Component({
    selector: 'select-test',
    template: '<ng-select></ng-select>'
})
class TestSelectEmptyComponent { }

@Component({
    selector: 'select-active-test',
    template: '<ng-select [items]="items"></ng-select>'
})
class TestSelectActiveComponent implements OnInit {
    formGroup: FormGroup;
    select: FormControl;
    items: Array<SelectItem> = new Array<SelectItem>();
    selected: Array<SelectItem>;
    /**
     *
     */
    constructor(private builder: FormBuilder) {
    }

    public ngOnInit(): void {
        this.select = this.builder.control(null, Validators.required);
        this.formGroup = this.builder.group({
            'select': this.select
        });

        setTimeout(() => {
            this.items.push(new SelectItem({ id: 1, text: 'Item 1' }));
            this.items.push(new SelectItem({ id: 2, text: 'Item 2' }));
            this.items.push(new SelectItem({ id: 3, text: 'Item 3' }));

            this.selected = [this.items[0]];
        });
    }
}

@Component({
    selector: 'select-active-test',
    template: '<ng-select [items]="items"></ng-select>'
})
class TestSelectAsyncComponent implements OnInit {
    items: Array<SelectItem>;

    constructor(private builder: FormBuilder) {
    }

    public onFilterChanged(filterText: string): void{
        setTimeout(() => {
            this.items = [
                new SelectItem({ id: 1, text: '***' + filterText + '***'}),
                new SelectItem({ id: 2, text: '***' + filterText + '***'}),
                new SelectItem({ id: 3, text: '***' + filterText + '***'})
            ];
        });
    }
}


demo/src/app/app.module.ts
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserModule } from '@angular/platform-browser';
import { CommonModule } from '@angular/common';
import { TabsModule, ButtonsModule } from 'ng2-bootstrap';

import { SelectModule } from 'ng2-select-hoveytech';
import { AppComponent } from './app.component';
import { SelectSectionComponent } from './components/select-section';
import { ChildrenDemoComponent } from './components/select/children-demo';
import { MultipleDemoComponent } from './components/select/multiple-demo';
import { RichDemoComponent } from './components/select/rich-demo';
import { SingleDemoComponent } from './components/select/single-demo';
import { SampleSectionComponent } from './components/sample-section.component';

@NgModule({
  declarations: [
    AppComponent,
    SampleSectionComponent,
    SelectSectionComponent,
    ChildrenDemoComponent,
    MultipleDemoComponent,
    RichDemoComponent,
    SingleDemoComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    ReactiveFormsModule,
    SelectModule,
    TabsModule.forRoot(),
    ButtonsModule.forRoot(),
    CommonModule
  ],
  bootstrap: [AppComponent]
})

export class AppModule {
}


demo/src/app/components/select/rich-demo.ts
import { Component, OnInit, ViewEncapsulation } from '@angular/core';

const COLORS = [
  {'name': 'Blue 10', 'hex': '#C0E6FF'},
  {'name': 'Blue 20', 'hex': '#7CC7FF'},
  {'name': 'Blue 30', 'hex': '#5AAAFA'},
  {'name': 'Blue 40', 'hex': '#5596E6'},
  {'name': 'Blue 50', 'hex': '#4178BE'},
  {'name': 'Blue 60', 'hex': '#325C80'},
  {'name': 'Blue 70', 'hex': '#264A60'},
  {'name': 'Blue 80', 'hex': '#1D3649'},
  {'name': 'Blue 90', 'hex': '#152935'},
  {'name': 'Blue 100', 'hex': '#010205'},
  {'name': 'Green 10', 'hex': '#C8F08F'},
  {'name': 'Green 20', 'hex': '#B4E051'},
  {'name': 'Green 30', 'hex': '#8CD211'},
  {'name': 'Green 40', 'hex': '#5AA700'},
  {'name': 'Green 50', 'hex': '#4B8400'},
  {'name': 'Green 60', 'hex': '#2D660A'},
  {'name': 'Green 70', 'hex': '#144D14'},
  {'name': 'Green 80', 'hex': '#0A3C02'},
  {'name': 'Green 90', 'hex': '#0C2808'},
  {'name': 'Green 100', 'hex': '#010200'},
  {'name': 'Red 10', 'hex': '#FFD2DD'},
  {'name': 'Red 20', 'hex': '#FFA5B4'},
  {'name': 'Red 30', 'hex': '#FF7D87'},
  {'name': 'Red 40', 'hex': '#FF5050'},
  {'name': 'Red 50', 'hex': '#E71D32'},
  {'name': 'Red 60', 'hex': '#AD1625'},
  {'name': 'Red 70', 'hex': '#8C101C'},
  {'name': 'Red 80', 'hex': '#6E0A1E'},
  {'name': 'Red 90', 'hex': '#4C0A17'},
  {'name': 'Red 100', 'hex': '#040001'},
  {'name': 'Yellow 10', 'hex': '#FDE876'},
  {'name': 'Yellow 20', 'hex': '#FDD600'},
  {'name': 'Yellow 30', 'hex': '#EFC100'},
  {'name': 'Yellow 40', 'hex': '#BE9B00'},
  {'name': 'Yellow 50', 'hex': '#8C7300'},
  {'name': 'Yellow 60', 'hex': '#735F00'},
  {'name': 'Yellow 70', 'hex': '#574A00'},
  {'name': 'Yellow 80', 'hex': '#3C3200'},
  {'name': 'Yellow 90', 'hex': '#281E00'},
  {'name': 'Yellow 100', 'hex': '#020100'}
];

@Component({
  selector: 'rich-demo',
  templateUrl: './rich-demo.html',
  styles: [`colorbox,.colorbox { display:inline-block; height:14px; width:14px;margin-right:4px; border:1px solid #000;}`],
  encapsulation: ViewEncapsulation.None  // Enable dynamic HTML styles
})
export class RichDemoComponent implements OnInit {
  public value:any = {};
  public disabled:boolean = false;
  public items:Array<any> = [];

  public ngOnInit():any {
    COLORS.forEach((color:{name:string, hex:string}) => {
      this.items.push({
        id: color.hex,
        text: `<colorbox style='background-color:${color.hex};'></colorbox>${color.name} (${color.hex})`
      });
    });
  }

  public toggleSelectDisabledState() {
    this.disabled = !this.disabled;
  }

  public selected(value:any):void {
    console.log('Selected value is: ', value);
  }

  public removed(value:any):void {
    console.log('Removed value is: ', value);
  }

  public typed(value:any):void {
    console.log('New search input: ', value);
  }

  public refreshValue(value:any):void {
    this.value = value;
  }
}


demo/src/app/components/select/single-demo.ts
import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, FormControl, Validators } from '@angular/forms';

@Component({
  selector: 'single-demo',
  templateUrl: './single-demo.html'
})
export class SingleDemoComponent implements OnInit {
  public items: Array<string> = ['Amsterdam', 'Antwerp', 'Athens', 'Barcelona',
    'Berlin', 'Birmingham', 'Bradford', 'Bremen', 'Brussels', 'Bucharest',
    'Budapest', 'Cologne', 'Copenhagen', 'Dortmund', 'Dresden', 'Dublin',
    'Düsseldorf', 'Essen', 'Frankfurt', 'Genoa', 'Glasgow', 'Gothenburg',
    'Hamburg', 'Hannover', 'Helsinki', 'Kraków', 'Leeds', 'Leipzig', 'Lisbon',
    'London', 'Madrid', 'Manchester', 'Marseille', 'Milan', 'Munich', 'Málaga',
    'Naples', 'Palermo', 'Paris', 'Poznań', 'Prague', 'Riga', 'Rome',
    'Rotterdam', 'Seville', 'Sheffield', 'Sofia', 'Stockholm', 'Stuttgart',
    'The Hague', 'Turin', 'Valencia', 'Vienna', 'Vilnius', 'Warsaw', 'Wrocław',
    'Zagreb', 'Zaragoza', 'Łódź'];
  public active: Array<string>;
  public value: any = {};
  public disabled: boolean = false;
  public formGroup: FormGroup;
  public selectControl: FormControl;

  constructor(private formBuilder: FormBuilder) {
    this.active = null;
  }

  public ngOnInit(): void {
    this.selectControl = this.formBuilder.control(null, Validators.required)
    this.formGroup = this.formBuilder.group({
      selectControl: this.selectControl
    });

    setTimeout(() => {
        this.active = [this.items[0]];
    });
  }

  public toggleSelectDisabledState() {
    this.disabled = !this.disabled;
  }

  public selected(value: any): void {
    console.log('Selected value is: ', value);
  }

  public removed(value: any): void {
    console.log('Removed value is: ', value);
  }

  public typed(value: any): void {
    console.log('New search input: ', value);
  }

  public refreshValue(value: any): void {
    this.value = value;
  }
}


demo/src/app/components/select/multiple-demo.ts
import { Component } from '@angular/core';

@Component({
  selector: 'multiple-demo',
  templateUrl: './multiple-demo.html'
})
export class MultipleDemoComponent {
  public items:Array<string> = ['Amsterdam', 'Antwerp', 'Athens', 'Barcelona',
    'Berlin', 'Birmingham', 'Bradford', 'Bremen', 'Brussels', 'Bucharest',
    'Budapest', 'Cologne', 'Copenhagen', 'Dortmund', 'Dresden', 'Dublin', 'Düsseldorf',
    'Essen', 'Frankfurt', 'Genoa', 'Glasgow', 'Gothenburg', 'Hamburg', 'Hannover',
    'Helsinki', 'Leeds', 'Leipzig', 'Lisbon', 'Łódź', 'London', 'Kraków', 'Madrid',
    'Málaga', 'Manchester', 'Marseille', 'Milan', 'Munich', 'Naples', 'Palermo',
    'Paris', 'Poznań', 'Prague', 'Riga', 'Rome', 'Rotterdam', 'Seville', 'Sheffield',
    'Sofia', 'Stockholm', 'Stuttgart', 'The Hague', 'Turin', 'Valencia', 'Vienna',
    'Vilnius', 'Warsaw', 'Wrocław', 'Zagreb', 'Zaragoza'];

  public value:any = ['Athens'];
  public disabled:boolean = false;

  public toggleSelectDisabledState() {
    this.disabled = !this.disabled;
  }

  public selected(value:any):void {
    console.log('Selected value is: ', value);
  }

  public removed(value:any):void {
    console.log('Removed value is: ', value);
  }

  public refreshValue(value:any):void {
    this.value = value;
  }

  public itemsToString(value:Array<any> = []):string {
    return value
      .map((item:any) => {
        return item.text;
      }).join(',');
  }
}


demo/src/app/components/select/children-demo.ts
import { Component } from '@angular/core';

@Component({
  selector: 'children-demo',
  templateUrl: './children-demo.html'
})
export class ChildrenDemoComponent {
  public items:Array<any> = [
    {
      text: 'Austria',
      children: [
        {id: 54, text: 'Vienna'}
      ]
    },
    {
      text: 'Belgium',
      children: [
        {id: 2, text: 'Antwerp'},
        {id: 9, text: 'Brussels'}
      ]
    },
    {
      text: 'Bulgaria',
      children: [
        {id: 48, text: 'Sofia'}
      ]
    },
    {
      text: 'Croatia',
      children: [
        {id: 58, text: 'Zagreb'}
      ]
    },
    {
      text: 'Czech Republic',
      children: [
        {id: 42, text: 'Prague'}
      ]
    },
    {
      text: 'Denmark',
      children: [
        {id: 13, text: 'Copenhagen'}
      ]
    },
    {
      text: 'England',
      children: [
        {id: 6, text: 'Birmingham'},
        {id: 7, text: 'Bradford'},
        {id: 26, text: 'Leeds'},
        {id: 30, text: 'London'},
        {id: 34, text: 'Manchester'},
        {id: 47, text: 'Sheffield'}
      ]
    },
    {
      text: 'Finland',
      children: [
        {id: 25, text: 'Helsinki'}
      ]
    },
    {
      text: 'France',
      children: [
        {id: 35, text: 'Marseille'},
        {id: 40, text: 'Paris'}
      ]
    },
    {
      text: 'Germany',
      children: [
        {id: 5, text: 'Berlin'},
        {id: 8, text: 'Bremen'},
        {id: 12, text: 'Cologne'},
        {id: 14, text: 'Dortmund'},
        {id: 15, text: 'Dresden'},
        {id: 17, text: 'Düsseldorf'},
        {id: 18, text: 'Essen'},
        {id: 19, text: 'Frankfurt'},
        {id: 23, text: 'Hamburg'},
        {id: 24, text: 'Hannover'},
        {id: 27, text: 'Leipzig'},
        {id: 37, text: 'Munich'},
        {id: 50, text: 'Stuttgart'}
      ]
    },
    {
      text: 'Greece',
      children: [
        {id: 3, text: 'Athens'}
      ]
    },
    {
      text: 'Hungary',
      children: [
        {id: 11, text: 'Budapest'}
      ]
    },
    {
      text: 'Ireland',
      children: [
        {id: 16, text: 'Dublin'}
      ]
    },
    {
      text: 'Italy',
      children: [
        {id: 20, text: 'Genoa'},
        {id: 36, text: 'Milan'},
        {id: 38, text: 'Naples'},
        {id: 39, text: 'Palermo'},
        {id: 44, text: 'Rome'},
        {id: 52, text: 'Turin'}
      ]
    },
    {
      text: 'Latvia',
      children: [
        {id: 43, text: 'Riga'}
      ]
    },
    {
      text: 'Lithuania',
      children: [
        {id: 55, text: 'Vilnius'}
      ]
    },
    {
      text: 'Netherlands',
      children: [
        {id: 1, text: 'Amsterdam'},
        {id: 45, text: 'Rotterdam'},
        {id: 51, text: 'The Hague'}
      ]
    },
    {
      text: 'Poland',
      children: [
        {id: 29, text: 'Łódź'},
        {id: 31, text: 'Kraków'},
        {id: 41, text: 'Poznań'},
        {id: 56, text: 'Warsaw'},
        {id: 57, text: 'Wrocław'}
      ]
    },
    {
      text: 'Portugal',
      children: [
        {id: 28, text: 'Lisbon'}
      ]
    },
    {
      text: 'Romania',
      children: [
        {id: 10, text: 'Bucharest'}
      ]
    },
    {
      text: 'Scotland',
      children: [
        {id: 21, text: 'Glasgow'}
      ]
    },
    {
      text: 'Spain',
      children: [
        {id: 4, text: 'Barcelona'},
        {id: 32, text: 'Madrid'},
        {id: 33, text: 'Málaga'},
        {id: 46, text: 'Seville'},
        {id: 53, text: 'Valencia'},
        {id: 59, text: 'Zaragoza'}
      ]
    },
    {
      text: 'Sweden',
      children: [
        {id: 22, text: 'Gothenburg'},
        {id: 49, text: 'Stockholm'}
      ]
    }
  ];
  public value:any = {};
  public disabled:boolean = false;

  public toggleSelectDisabledState() {
    this.disabled = !this.disabled;
  }

  public selected(value:any):void {
    console.log('Selected value is: ', value);
  }

  public removed(value:any):void {
    console.log('Removed value is: ', value);
  }

  public refreshValue(value:any):void {
    this.value = value;
  }
}
